sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -507,12 +508,13 @@ def _fit_stochastic(self, X, y, activations, deltas, coef_grads,
         try:
             for it in range(self.max_iter):
                 if self.shuffle:
-                    X, y = shuffle(X, y, random_state=self._random_state)
+                    idx = shuffle(idx, random_state=self._random_state)
                 accumulated_loss = 0.0
                 for batch_slice in gen_batches(n_samples, batch_size):
-                    activations[0] = X[batch_slice]
+                    activations[0] = X[idx[batch_slice]]
```

Comment:
This will not work with COO sparse matrix. We either need to remove support for 'COO' matrices and convert them into 'CSC' or 'CSR' during the validation. Then, I think that you should use `safe_indexing`:

```python
from ..utils import safe_indexing
safe_indexing(X, idx[batch_slice])
```